cmake/CheckHeaders.cmake

```diff
@@ -0,0 +1,113 @@
+# a poor's alternative of iwyu
+#
+# Header files should be self-contained. In general, the source files should
+# not have to adhere to special conditions to include them. For instance,
+# they don't need to include other header files for using a header file, or
+# to define certain macro(s) for using it. But the macros are allowed to be
+# used to its behavior though.
+#
+# include-what-you-use is able to check if all symbols used by a certain header
+# are defined by the headers *directly* included by that header by parsing the
+# header source file using libclang. But we don't always have iwyu installed.
+# Instead of relying on a tool, a new CMake function named
+# "seastar_check_self_contained()" is created to perform a minimal check on the
+# specified set of header files by compiling each of them.
+#
+# Please note, "seastar_check_self_contained()" does not validate the
+# "self-contained" criteria defined above, it just check if the header file
+# *contains* the declarations of the symbols so that any source file including
+# it can be compiled as well. The check performed by the CMake function allows
+# the indirect inclusion of the used symbols. For instance, if "juno.h"
+# references a symbol named "Jupiter", which is declared in "jupiter.h". So,
+# strictly speaking, "juno.h" should include "jupiter.h" for accessing this
+# symbol. But "juno.h" happens to include "solar.h", which in turn includes
+# "jupiter.h". So "seastar_check_self_contained()" accepts "juno.h", while a
+# tool like iwyu would complain at seeing it.
+#
+# If you have include-what-you-use installed, it'd be advisable to configure
+# the building system with `-DCMAKE_CXX_INCLUDE_WHAT_YOU_USE=include-what-you-use`,
+# and build the tree.
+
+function (seastar_check_self_contained library)
+  cmake_parse_arguments (
+    parsed_args
+    ""
+    ""
+    "EXCLUDE;INCLUDE"
+    ${ARGN})
+
+  get_target_property (sources ${library} SOURCES)
+  list (FILTER sources INCLUDE REGEX "${parsed_args_INCLUDE}")
+  list (FILTER sources EXCLUDE REGEX "${parsed_args_EXCLUDE}")
+  foreach (fn ${sources})
```

Comment:
I don't understand what I'm seeing here. Is this an imperative function which loops over all header files and does the check on all of them, every time? I really hope the answer is (if this is the case, it can be a shell script or python or what you prefer, it doesn't need to be this limited cmake syntax).

Or does this build make rules - so that if I modify only one header file, only this header file **and what includes it** is checked and not all the headers are rechecked? (please check the bold part above in particular, holds).

Reply:
> I don't understand what I'm seeing here. Is this an imperative function which loops over all header files and does the check on all of them, every time? I really hope the answer is (if this is the case, it can be a shell script or python or what you prefer, it doesn't need to be this limited cmake syntax).

no, it is not.

> 
> Or does this build make rules - so that if I modify only one header file, only this header file **and what includes it** is checked and not all the headers are rechecked? (please check the bold part above in particular, holds).

yes. that how it works -- by compile it. did you get a chance to read the commit message of this commit? if not, it should help.

Reply:
> I don't understand how the commit message answers my question.

sorry for the confusion. i will try harder.

> 
> Anyway, because my cmake skills are virtually non-existent, I still don't understand what happens if I modify just one header file - will cmake copy just this one header file to a ".cc" file, and then only recompile the changed .cc file?

i think your understanding is quite right. the compiler "checks" if the header file is self-contained by compiling it. if the .cc file does not compile, then the original .hh file is not self-contained.

> And I guess it will also recompile other fake-".cc" files which included changed header files? It is very possible you did everything right - I just wanted to make sure because I can't understand the implementation :-(

i think so. please note the generated .cc file still includes the original .hh files.

Reply:
@nyh i updated the commit message with more details. HTH.